### PR TITLE
Fix/28925: Anonymous user can edit profile

### PR DESCRIPTION
**Pull request by DylanDylann**
### Details

Public room as anonymous user can able to edit profile
### Fixed Issues

$ https://github.com/Expensify/App/issues/28925
PROPOSAL: https://github.com/Expensify/App/issues/28925#issuecomment-1749125001

### Tests

1. Go to a public room as an anonymous user like https://staging.new.expensify.com/r/207591744844000
2. Click on the search icon.
3. Close the login menu.
4. Enter url in chrome /settings/profile, /settings/profile/display-name
5. Verify that the login modal is displayed
- [x] Verify that no errors appear in the JS console

### QA Steps

1. Go to a public room as an anonymous user like https://staging.new.expensify.com/r/207591744844000
2. Click on the search icon.
3. Close the login menu.
4. Enter url in chrome /settings/profile, /settings/profile/display-name
5. Verify that the login modal is displayed
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

There is a problem when testing with mobile native platform: Go to the public room with an anonymous user encountered the bellow error: (Which existed in the latest staging as well)

Screenshot from 2023-10-11 16-47-07
</details>

<details>
<summary>Android: mWeb Chrome</summary>

[Screencast from 11-10-2023 09:21:00.webm](https://github.com/Expensify/App/assets/141406735/929950e4-2c01-4ce1-a352-b85593c119fc)

</details>

<details>
<summary>iOS: Native</summary>

There is a problem when testing with mobile native platform: Go to the public room with an anonymous user encountered the bellow error: (Which existed in the latest staging as well)

Screenshot from 2023-10-11 16-47-07
</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/141406735/3f1b909c-98f6-434a-a426-97acfa901791

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/141406735/d0104463-331c-4d38-b227-0ccf606c2c93

</details>

<details>
<summary>MacOS: Desktop</summary>

Cannot go to the public room with an anonymous user on desktop
</details>

### CODE REVIEW

**Comment by DylanDylann** [comment]
1. There is a problem when testing with mobile native platform: Go to the public room with an anonymous user encountered the bellow error: (Which existed in the latest staging as well)

Screenshot from 2023-10-11 16-47-07

2. Also I cannot go to the public room with an anonymous user on desktop.

So currently, I am missing the mobile native and desktop screenshots

**Comment by DylanDylann** [comment]
@burczu  please help review this PR when you have a chance. Thanks

**Comment by burczu** [comment]
@DylanDylann In terms native and desktop, I think we can skip testing them here, so please check all the checkboxes and leave a suitable comments in screenshots/videos section of native environments.

**Comment by burczu** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>



</details>

<details>
<summary>Mobile Web - Chrome</summary>



</details>

<details>
<summary>Mobile Web - Safari</summary>



</details>

<details>
<summary>Desktop</summary>



</details>

<details>
<summary>iOS</summary>



</details>

<details>
<summary>Android</summary>



</details>

**Comment by burczu** [comment]
@DylanDylann I've encountered strange behaviour while testing this PR. Please take a look at the video below:

https://github.com/Expensify/App/assets/7771498/61edc9bd-f9db-4944-a00f-5030b5233d74

I think this should be handled here too, because it may case a regression.

**Comment by DylanDylann** [comment]
> @DylanDylann I've encountered strange behaviour while testing this PR. Please take a look at the video below:
> 
>  Screen.Recording.2023-10-13.at.07.42.57.mov 
> I think this should be handled here too, because it may case a regression.

Do you have any suggestions about the expected behavior?

**Comment by burczu** [comment]
@DylanDylann Two options: clicking back should do nothing or removing this back button.

**Comment by burczu** [comment]
@DylanDylann Actually - wait :) I think this back button should actually just close the sidebar. What do you think?

**Comment by DylanDylann** [comment]
@burczu yeah. So i think we just need to close the side bar when clicking on back button

**Comment by DylanDylann** [comment]
@burczu just updated the PR and here is the result:

[Screencast from 13-10-2023 13:40:40.webm](https://github.com/Expensify/App/assets/141406735/fd8279d5-4f95-4185-af1d-bf01093c18a1)

**Comment by DylanDylann** [comment]
@burczu please help review this PR when you have a chance. Thanks

**Comment by burczu** [comment]
@DylanDylann Thanks for addressing the issue with back button. Unfortunately I've encountered the same issue (second sidebar shows up for a second) event when I try to open `/settings/profile` or `/settings/profile/display-name`...

https://github.com/Expensify/App/assets/7771498/8dc70a64-b927-46f1-b219-0dceb6013557

**Comment by DylanDylann** [comment]
@burczu I think it is an edge case because I can just handle the ```onStateChange``` event

**Comment by DylanDylann** [comment]
> @DylanDylann Thanks for addressing the issue with back button. Unfortunately I've encountered the same issue (second sidebar shows up for a second) event when I try to open `/settings/profile` or `/settings/profile/display-name`...
> 
>  Screen.Recording.2023-10-16.at.14.21.26.mov

@stitesExpensify should we consider it as an edge case?

**Comment by burczu** [comment]
@DylanDylann I don't think we should consider it as an edge case - there is a glitch and I bet it will be raised as a bug in the future, so I think we should try to find a solution for this. Sorry for the delayed answer, I had a lot other thinks to work on. I'll think about the possible solution here later today.

**Comment by DylanDylann** [comment]
@burczu please help review the latest change. Here is the result:

[Screencast from 19-10-2023 01:11:51.webm](https://github.com/Expensify/App/assets/141406735/955f0b35-71a5-4dad-92f0-0de4138b72fd)

**Comment by burczu** [comment]
@DylanDylann I can confirm the glitch is fixed now. Well done! I'll finish the review shortly.

**Comment by burczu** [comment]
@DylanDylann Just checked you latest changes and it didn't fixed it, it's even worse I would say... Please check the below video (especially the very end of it):

https://github.com/Expensify/App/assets/7771498/eb91ec83-49fc-4bca-bca2-1e3acf8c29c8

**Comment by DylanDylann** [comment]
@burczu let me check again

**Comment by DylanDylann** [comment]
@burczu Updated the PR. The new solution will handle two case: 
1. User opens app by deeplink
2. User clicks on the link that has the same origin as our app  in any message on report screen

Here is the result:

[Screencast from 20-10-2023 14:59:07.webm](https://github.com/Expensify/App/assets/141406735/b495e405-f688-43b9-b071-f70b7718a2ab)

**Comment by burczu** [comment]
@DylanDylann Yes, I'll get back to this PR today. Could you please resolve conflicts in the meantime? I guess sync with `main` is necessary.

**Comment by burczu** [comment]
Hey @DylanDylann I've checked the new solution and I think it's the step in a good direction - I still can see the `/settings/profile` modal but for a shorted amount of time. I think the problem here is with this navigation - it looks like it starts before the redirection to the login modal page is made. Maybe we could somehow postpone it a little bit?

I tend to think I could approve it and see what the Expensify engineer thinks about it.

**Comment by DylanDylann** [comment]
@burczu I`ve fixed the conflict but the latest change in the main branch makes the app crash so I cannot test. Will ping you when it is fixed https://github.com/Expensify/App/pull/30228#issuecomment-1779203025

**Comment by DylanDylann** [comment]
> 1. r/207591744844000

@burczu the main is fixed, just tested and here is the result:

[Screencast from 26-10-2023 10:11:23.webm](https://github.com/Expensify/App/assets/141406735/e4254c82-2c58-4b80-b143-c97b0bc0f083)

**Review by burczu — APPROVED** [review]
Code LGTM

**Comment by burczu** [comment]
@stitesExpensify I've approved this PR, but there is one thing that concerns me. Please see the below video:

https://github.com/Expensify/App/assets/7771498/bb8c48b3-026f-430a-9aa1-8deaf752e2ef

When we type `/settings/profile` or `/settings/profile/display-name` the page is correctly redirected to the Login modal, bot for milliseconds we see the settings page, until the sliding animation of the login sidebar ends - I'm not sure if this acceptable, so please decide about it.

**Comment by DylanDylann** [comment]
@stitesExpensify please help review this PR when you have a chance

**Comment by DylanDylann** [comment]
@stitesExpensify friendly bump in case you miss this one

**Review by stitesExpensify — APPROVED** [review]

File: src/pages/signin/SignInModal.js [hunk 1]
```diff
@@ -24,7 +24,7 @@ function SignInModal() {
             shouldEnableMaxHeight
             testID={SignInModal.displayName}
         >
-            <HeaderWithBackButton />
+            <HeaderWithBackButton onBackButtonPress={Navigation.dismissModal} />
```
Comment (s77rt) [inline src/pages/signin/SignInModal.js:27]: This caused inconsistency between browser back button and modal back button. (Coming from https://github.com/Expensify/App/issues/33417)

File: src/libs/actions/Session/index.ts [hunk 2]
```diff
@@ -871,6 +871,33 @@ function waitForUserSignIn(): Promise<boolean> {
     });
 }
 
+/**
+ * check if the route can be accessed by anonymous user
+ *
+ * @param {string} route
+ */
+
+const canAccessRouteByAnonymousUser = (route: string) => {
```
Comment (flodnv) [inline src/libs/actions/Session/index.ts:880]: `canAnonymousUserAccessRoute` would sound better

File: src/libs/actions/Session/index.ts [hunk 3]
```diff
@@ -871,6 +871,33 @@ function waitForUserSignIn(): Promise<boolean> {
...
+    }
+    const parsedReportRouteParams = ReportUtils.parseReportRouteParams(route);
+    let routeRemovedReportId = route;
+    if ((parsedReportRouteParams as {reportID: string})?.reportID) {
+        routeRemovedReportId = route.replace((parsedReportRouteParams as {reportID: string})?.reportID, ':reportID');
+    }
+    if (route.startsWith('/')) {
+        routeRemovedReportId = routeRemovedReportId.slice(1);
+    }
+    const routesCanAccessByAnonymousUser = [ROUTES.SIGN_IN_MODAL, ROUTES.REPORT_WITH_ID_DETAILS.route, ROUTES.REPORT_WITH_ID_DETAILS_SHARE_CODE.route];
```
Comment (flodnv) [inline src/libs/actions/Session/index.ts:893]: `routesAccessibleByAnonymousUser` would sound better

File: src/libs/actions/Session/index.ts [hunk 4]
```diff
@@ -871,6 +871,33 @@ function waitForUserSignIn(): Promise<boolean> {
     });
 }
 
+/**
+ * check if the route can be accessed by anonymous user
```
Comment (flodnv) [inline src/libs/actions/Session/index.ts:875]: Updating this comment (or adding a comment in the function) to say why an anonymous user can only access different routes would be helpful for future programmers like me who don't have context on why this exists (in general, comments should be used to explain _why_ the code is like that, not _what_ the code is obviously doing)

File: src/libs/actions/Session/index.ts (hunk 3 above)
Comment (flodnv) [reply src/libs/actions/Session/index.ts:893]: Fixed here: https://github.com/Expensify/App/pull/37952

**Pull request merged**